### PR TITLE
Add `.apply_layout()`method  to `Pauli` class

### PR DESCRIPTION
**Pull request by ElePT**
### Summary
This PR adds an `apply_layout` method to the `Pauli` class, following what was done for `SparsePauliOp` in https://github.com/Qiskit/qiskit/pull/10947. This was the agreed action to partially solve #11824 in the short term until a more permanent/extensible solution is designed.

### CODE REVIEW

File: qiskit/quantum_info/operators/symplectic/pauli.py [hunk 1]
```diff
@@ -697,6 +698,49 @@ def _from_circuit(cls, instr):
...
+            layout: Either a :class:`~.TranspileLayout`, a list of integers or None.
+                    If both layout and num_qubits are none, a copy of the operator is
+                    returned.
+            num_qubits: The number of qubits to expand the operator to. If not
+                provided then if ``layout`` is a :class:`~.TranspileLayout` the
+                number of the transpiler output circuit qubits will be used by
+                default. If ``layout`` is a list of integers the permutation
+                specified will be applied without any expansion. If layout is
+                None, the operator will be expanded to the given number of qubits.
+
```
Comment (mtreinish) [inline qiskit/quantum_info/operators/symplectic/pauli.py:716]:
```suggestion
```

File: qiskit/quantum_info/operators/symplectic/pauli.py [hunk 2]
```diff
@@ -697,6 +698,49 @@ def _from_circuit(cls, instr):
...
+                raise QiskitError(
+                    f"The input num_qubits is too small, a {num_qubits} qubit layout cannot be "
+                    f"applied to a {n_qubits} qubit operator"
+                )
+            n_qubits = num_qubits
+        if layout is not None and any(x >= n_qubits for x in layout):
+            raise QiskitError("Provided layout contains indices outside the number of qubits.")
+        if layout is None:
+            layout = list(range(self.num_qubits))
+        new_op = type(self)("I" * n_qubits)
```
Comment (mtreinish) [inline qiskit/quantum_info/operators/symplectic/pauli.py:740]: I realize I did the same thing in #10947 but I'm wondering if there is a more efficient constructor here that doesn't require creating a giant string like this.

**Review by mtreinish — COMMENTED** [review]
This LGTM, just a whitespace nit inline and a small question. But otherwise this is straightforward as it's just the same method as `SparsePauliOp`.

File: qiskit/quantum_info/operators/symplectic/pauli.py [hunk 3]
```diff
@@ -697,6 +698,49 @@ def _from_circuit(cls, instr):
...
+            if num_qubits < n_qubits:
+                raise QiskitError(
+                    f"The input num_qubits is too small, a {num_qubits} qubit layout cannot be "
+                    f"applied to a {n_qubits} qubit operator"
+                )
+            n_qubits = num_qubits
+        if layout is not None and any(x >= n_qubits for x in layout):
+            raise QiskitError("Provided layout contains indices outside the number of qubits.")
+        if layout is None:
+            layout = list(range(self.num_qubits))
```
Comment (MaitreyPrajapati) [inline qiskit/quantum_info/operators/symplectic/pauli.py]: Can these two conditions be merged as following?

```
if layout is None:
    layout = list(range(self.num_qubits))
elif any(x >= n_qubits for x in layout):
    raise QiskitError("Provided layout contains indices outside the number of qubits.")
```

File: qiskit/quantum_info/operators/symplectic/pauli.py (hunk 2 above)
Comment (ElePT) [reply qiskit/quantum_info/operators/symplectic/pauli.py:740]: Hmmm, as far as I know, there is no good way to "sparsely" define long operators with only a few non-identity terms, in fact, I think that this is exactly what this issue describes (the word "layout" has a different meaning here: https://github.com/Qiskit/qiskit/issues/11891).

File: qiskit/quantum_info/operators/symplectic/pauli.py (hunk 3 above)
Comment (ElePT) [reply qiskit/quantum_info/operators/symplectic/pauli.py]: Sure, applied in [9778aa6](https://github.com/Qiskit/qiskit/pull/12066/commits/9778aa6e43bad015512a86a9543cc7e6f7c608a1)

Comment (MaitreyPrajapati) [reply qiskit/quantum_info/operators/symplectic/pauli.py]: Thanks. LGTM otherwise

File: qiskit/quantum_info/operators/symplectic/pauli.py (hunk 2 above)
Comment (jakelishman) [reply qiskit/quantum_info/operators/symplectic/pauli.py:740]: Well, you can directly generate the symplectic Boolean matrices, but fundamentally, `Pauli` and `SparsePauliOp` aren't sparse over qubit number - the latter is "sparse" over the full Pauli vector.

Max's issue is about that sort of thing, yeah, although practically I don't think it's something we'll be able to put directly into `Pauli` and `SparsePauliOp` - I think it's something we'd do as new QI classes, probably directly in Rust space.  This it the kind of thing that Paul's been interested in before as well, and similar to how OpenFermion and friends represent operators on huge spaces.  The primitives want that as well.

File: test/python/quantum_info/operators/symplectic/test_pauli.py [hunk 4]
```diff
@@ -495,6 +501,87 @@ def test_circuit_with_bit(self):
...
+        transpiled_psi = transpile(psi, backend, optimization_level=3, seed_transpiler=12345)
+        permuted_op = op.apply_layout(transpiled_psi.layout)
+        identity_op = Pauli("I" * 7)
+        initial_layout = transpiled_psi.layout.initial_index_layout(filter_ancillas=True)
+        final_layout = transpiled_psi.layout.routing_permutation()
+        qargs = [final_layout[x] for x in initial_layout]
+        expected_op = identity_op.compose(op, qargs=qargs)
+        self.assertNotEqual(op, permuted_op)
+        self.assertEqual(permuted_op, expected_op)
+
```
Comment (jakelishman) [inline test/python/quantum_info/operators/symplectic/test_pauli.py:518]: Just for completeness' sake, could you add a test that `SparsePauliOp(Pauli(p).apply_layout(layout))` produces the same result as `SparsePauliop(Pauli(p)).apply_layout(layout)`?  Just a consistency test, to enforce that we don't forget one or the other while updating.

**Review by jakelishman — COMMENTED** [review]
The tests look right to me (and definitely something we should have), thanks for doing it!

File: test/python/quantum_info/operators/symplectic/test_pauli.py (hunk 4 above)
Comment (ElePT) [reply test/python/quantum_info/operators/symplectic/test_pauli.py:518]: I think that you were meaning to say `SparsePauliOp(Pauli(p).apply_layout(layout))` produces the same result as `Pauli(p).apply_layout(layout)`, right? This is what I added in [434a21c](https://github.com/Qiskit/qiskit/pull/12066/commits/434a21c791d124ed542632bedfbf4ef20e917472).

Comment (jakelishman) [reply test/python/quantum_info/operators/symplectic/test_pauli.py:518]: Yeah, that's fine too - you've ended up reducing the `SparsePauliOp` to a `PauliList` to check, whereas I was suggesting raising the `Pauli` to a `SparsePauliOp` to compare, but both test the same thing that I wanted to test.

Comment (ElePT) [reply test/python/quantum_info/operators/symplectic/test_pauli.py:518]: ahaa, I didn't notice the brackets. I can also do it the way you originally suggested (looks cleaner to compare the exact same type).

**Review by jakelishman — APPROVED** [review]
Thanks for the extra test!

**Pull request merged**